File: Problems/Trees/Lowest-Common-Ancestor-of-a-Binary-Search-Tree/Lowest-Common-Ancestor-of-a-Binary-Search-Tree.cs
using System;

// Using recursion - Time: O(h)

public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution
{
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q)
    {
        if (root.val > p.val && root.val > q.val)
        {
            return LowestCommonAncestor(root.left, p, q);
        }
        if (root.val < p.val && root.val < q.val)
        {
            return LowestCommonAncestor(root.right, p, q);
        }
        return root;
    }
}

class Program
{
    static void PrintTree(TreeNode root)
    {
        if (root == null)
        {
            Console.Write("null");
            return;
        }

        List<string> values = new List<string>();
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        while (queue.Count > 0)
        {
            TreeNode current = queue.Dequeue();
            if (current != null)
            {
                values.Add(current.val.ToString());
                queue.Enqueue(current.left);
                queue.Enqueue(current.right);
            }
            else
            {
                values.Add("null");
            }
        }

        Console.Write("[" + string.Join(", ", values) + "]");
    }
    static void Main()
    {
        TreeNode root = new TreeNode(6)
        {
            left = new TreeNode(2)
            {
                left = new TreeNode(0),
                right = new TreeNode(4)
                {
                    left = new TreeNode(3),
                    right = new TreeNode(5)
                }
            },
            right = new TreeNode(8)
            {
                left = new TreeNode(7),
                right = new TreeNode(9)
            }
        };

        TreeNode p = new TreeNode(2);
        TreeNode q = new TreeNode(8);

        Console.Write("Input: root = ");
        PrintTree(root);
        Console.WriteLine(", p = " + p.val + ", q = " + q.val);

        Solution sol = new Solution();
        TreeNode result = sol.LowestCommonAncestor(root, p, q);

        Console.WriteLine("Output: " + result.val);
    }
}

File: Problems/Trees/Lowest-Common-Ancestor-of-a-Binary-Search-Tree/Lowest-Common-Ancestor-of-a-Binary-Search-Tree.cpp
#include <iostream>
#include <vector>
#include <queue>

// Using recursion - Time: O(h)

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q)
    {
        if (root->val > p->val && root->val > q->val)
            return lowestCommonAncestor(root->left, p, q);
        if (root->val < p->val && root->val < q->val)
            return lowestCommonAncestor(root->right, p, q);
        return root;
    }
};

void printTree(TreeNode *root)
{
    if (!root)
    {
        std::cout << "[]" << std::endl;
        return;
    }

    std::queue<TreeNode *> queue;
    queue.push(root);

    std::cout << "[";

    while (!queue.empty())
    {
        int n = queue.size();

        for (int i = 0; i < n; i++)
        {
            TreeNode *node = queue.front();
            queue.pop();

            if (node)
            {
                std::cout << node->val << ", ";

                queue.push(node->left);
                queue.push(node->right);
            }
            else
            {
                if (!queue.empty())
                    std::cout << "null, ";
                else
                    std::cout << "null";
            }
        }
    }

    std::cout << "]";
}

int main()
{
    TreeNode *root = new TreeNode(6);
    root->left = new TreeNode(2);
    root->left->left = new TreeNode(0);
    root->left->right = new TreeNode(4);
    root->left->right->left = new TreeNode(3);
    root->left->right->right = new TreeNode(5);

    root->right = new TreeNode(8);
    root->right->left = new TreeNode(7);
    root->right->right = new TreeNode(9);

    TreeNode *p = new TreeNode(2);
    TreeNode *q = new TreeNode(8);

    std::cout << "Input: root = ";
    printTree(root);
    std::cout << ", p = " << p->val << ", q = " << q->val << std::endl;

    Solution sol;
    TreeNode *ans = sol.lowestCommonAncestor(root, p, q);

    std::cout << "Output: " << ans->val << std::endl;

    delete root;
    delete p;
    delete q;

    return 0;
}
